File: src/algorithms/graph/shortest-path/floyd-warshall/floyd-warshall.test.ts
import { describe, it, expect } from "vitest";

import { floydWarshall } from "./sources/floyd-warshall.ts?fn";

type WeightedAdjacencyList = Record<string, [string, number][]>;

describe("floydWarshall", () => {
  it("computes all-pairs shortest paths in a 4-node graph", () => {
    // Graph: A→B(3), A→D(-4), D→C(6), C→B(-5). No negative cycles.
    // Shortest A→B = min(3, -4+6-5) = -3 via A→D→C→B
    // Shortest A→C = -4+6 = 2 via A→D→C
    // Shortest A→D = -4 (direct)
    const adjacencyList: WeightedAdjacencyList = {
      A: [
        ["B", 3],
        ["D", -4],
      ],
      B: [],
      C: [["B", -5]],
      D: [["C", 6]],
    };
    const distances = floydWarshall(adjacencyList, ["A", "B", "C", "D"]);
    expect(distances["A"]!["A"]).toBe(0);
    expect(distances["A"]!["B"]).toBe(-3); // A→D(-4)→C(6)→B(-5) = -3
    expect(distances["A"]!["C"]).toBe(2); // A→D(-4)→C(6) = 2
    expect(distances["B"]!["B"]).toBe(0);
  });

  it("sets diagonal entries to zero", () => {
    const adjacencyList: WeightedAdjacencyList = {
      X: [["Y", 2]],
      Y: [["Z", 3]],
      Z: [],
    };
    const distances = floydWarshall(adjacencyList, ["X", "Y", "Z"]);
    expect(distances["X"]!["X"]).toBe(0);
    expect(distances["Y"]!["Y"]).toBe(0);
    expect(distances["Z"]!["Z"]).toBe(0);
  });

  it("returns Infinity for unreachable node pairs", () => {
    const adjacencyList: WeightedAdjacencyList = {
      A: [["B", 1]],
      B: [],
      C: [],
    };
    const distances = floydWarshall(adjacencyList, ["A", "B", "C"]);
    expect(distances["A"]!["C"]).toBe(Infinity);
    expect(distances["C"]!["A"]).toBe(Infinity);
  });

  it("handles a single-node graph", () => {
    const adjacencyList: WeightedAdjacencyList = { A: [] };
    const distances = floydWarshall(adjacencyList, ["A"]);
    expect(distances["A"]!["A"]).toBe(0);
  });

  it("finds shorter indirect paths over direct edges", () => {
    // A→C direct = 10, but A→B(1)→C(2) = 3
    const adjacencyList: WeightedAdjacencyList = {
      A: [
        ["B", 1],
        ["C", 10],
      ],
      B: [["C", 2]],
      C: [],
    };
    const distances = floydWarshall(adjacencyList, ["A", "B", "C"]);
    expect(distances["A"]!["C"]).toBe(3);
  });

  it("computes correct bidirectional distances in an undirected-style graph", () => {
    const adjacencyList: WeightedAdjacencyList = {
      A: [["B", 4]],
      B: [
        ["A", 4],
        ["C", 3],
      ],
      C: [["B", 3]],
    };
    const distances = floydWarshall(adjacencyList, ["A", "B", "C"]);
    expect(distances["A"]!["C"]).toBe(7);
    expect(distances["C"]!["A"]).toBe(7);
  });

  it("correctly handles negative edge weights without negative cycles", () => {
    const adjacencyList: WeightedAdjacencyList = {
      A: [["B", 5]],
      B: [["C", -2]],
      C: [],
    };
    const distances = floydWarshall(adjacencyList, ["A", "B", "C"]);
    expect(distances["A"]!["C"]).toBe(3);
    expect(distances["A"]!["B"]).toBe(5);
  });
});

File: src/algorithms/graph/shortest-path/floyd-warshall/__tests__/FloydWarshall_test.cpp
#include "../sources/FloydWarshall.cpp"
#include <cassert>
#include <iostream>
#include <limits>

int main() {
    // Test 1: 4-node all-pairs
    {
        WeightedAdjList adj = {
            {"A",{{"B",3},{"D",-4}}},{"B",{}},
            {"C",{{"B",-5}}},{"D",{{"C",6}}}
        };
        auto d = FloydWarshall::floydWarshall(adj, {"A","B","C","D"});
        assert(d.at("A").at("A") == 0);
        assert(d.at("A").at("B") == -3);
        assert(d.at("A").at("C") == 2);
        assert(d.at("B").at("B") == 0);
    }

    // Test 2: diagonal is zero
    {
        WeightedAdjList adj = {{"X",{{"Y",2}}},{"Y",{{"Z",3}}},{"Z",{}}};
        auto d = FloydWarshall::floydWarshall(adj, {"X","Y","Z"});
        assert(d.at("X").at("X") == 0);
        assert(d.at("Y").at("Y") == 0);
        assert(d.at("Z").at("Z") == 0);
    }

    // Test 3: unreachable pairs
    {
        WeightedAdjList adj = {{"A",{{"B",1}}},{"B",{}},{"C",{}}};
        auto d = FloydWarshall::floydWarshall(adj, {"A","B","C"});
        assert(d.at("A").at("C") == numeric_limits<int>::max());
        assert(d.at("C").at("A") == numeric_limits<int>::max());
    }

    // Test 4: single node
    {
        WeightedAdjList adj = {{"A",{}}};
        auto d = FloydWarshall::floydWarshall(adj, {"A"});
        assert(d.at("A").at("A") == 0);
    }

    // Test 5: shorter indirect path
    {
        WeightedAdjList adj = {{"A",{{"B",1},{"C",10}}},{"B",{{"C",2}}},{"C",{}}};
        auto d = FloydWarshall::floydWarshall(adj, {"A","B","C"});
        assert(d.at("A").at("C") == 3);
    }

    // Test 6: bidirectional
    {
        WeightedAdjList adj = {{"A",{{"B",4}}},{"B",{{"A",4},{"C",3}}},{"C",{{"B",3}}}};
        auto d = FloydWarshall::floydWarshall(adj, {"A","B","C"});
        assert(d.at("A").at("C") == 7);
        assert(d.at("C").at("A") == 7);
    }

    // Test 7: negative weights
    {
        WeightedAdjList adj = {{"A",{{"B",5}}},{"B",{{"C",-2}}},{"C",{}}};
        auto d = FloydWarshall::floydWarshall(adj, {"A","B","C"});
        assert(d.at("A").at("C") == 3);
        assert(d.at("A").at("B") == 5);
    }

    cout << "All tests passed!" << endl;
    return 0;
}
